### PR TITLE
feat: add custom pixel event

### PR DESCRIPTION
**Pull request by gutogama**
#### What problem is this solving?

Trigger custom pixel event for newsletter tracking purposes. Requested by Dito CRM.

#### How to test it?

A new message is being dispatched and can be checked adding an event listener of type 'message' to the Window

[Workspace](Link goes here!)

#### Screenshots or example usage:

```js
window.addEventListener('message', function(e) {
  if (e.data.eventName === 'vtex:newsletterSubscription') {
    console.log(e.data);
  }
})
```

### CODE REVIEW

File: react/Newsletter.tsx [hunk 1]
```diff
@@ -94,6 +95,12 @@ function Newsletter(props: PropsWithChildren<Props>) {
       return
     }
 
+    triggerPixelNewsletterEvent({
```
Comment (klzns) [inline react/Newsletter.tsx:98]: Use the hook usePixel, example:

1. https://github.com/vtex-apps/add-to-cart-button/blob/master/react/AddToCartButton.tsx#L12

2. https://github.com/vtex-apps/add-to-cart-button/blob/master/react/AddToCartButton.tsx#L120

3. https://github.com/vtex-apps/add-to-cart-button/blob/master/react/AddToCartButton.tsx#L192-L205

**Review by klzns — CHANGES_REQUESTED** [review]

File: react/Newsletter.tsx (hunk 1 above)
Comment (Mogueno) [reply react/Newsletter.tsx:98]: @klzns Sure!
I will assume this thread from now on.

Btw I am also waiting for another PR to be accepted to move on with this feature 
https://github.com/vtex-apps/pixel-app-template/pull/15

Comment (Mogueno) [reply react/Newsletter.tsx]: @klznsDone with the native hook as suggested, can you please review this part before I update the docs and remove the old implemetation files?

File: react/Newsletter.tsx [hunk 2]
```diff
@@ -1,124 +1,149 @@
-import React, { ComponentType, PropsWithChildren, FormEvent } from 'react'
-import { useCssHandles } from 'vtex.css-handles'
+import React, { ComponentType, PropsWithChildren, FormEvent } from "react";
```
Comment (victorhmp) [inline react/Newsletter.tsx:1]: The linting/formatting on this file is off. Please make sure to run `yarn lint --fix` in this file.

**Review by victorhmp — CHANGES_REQUESTED** [review]

File: react/Newsletter.tsx (hunk 2 above)
Comment (Mogueno) [reply react/Newsletter.tsx]: @victorhmp done!

File: react/interfaces/custom-pixel-event.interface.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+export interface CustomPixelEventInterface {
```
Comment (victorhmp) [inline react/interfaces/custom-pixel-event.interface.ts:1]: I guess this file is not needed anymore

File: react/modules/customPixelEvents.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import { CustomPixelEventInterface } from '../interfaces/custom-pixel-event.interface'
```
Comment (victorhmp) [inline react/modules/customPixelEvents.ts:1]: This file is also not needed

File: react/typings/vtex.pixel-manager.d.ts [hunk 5]
```diff
@@ -0,0 +1 @@
+declare module 'vtex.pixel-manager'
```
Comment (victorhmp) [inline react/typings/vtex.pixel-manager.d.ts:1]: This should also be unnecessary. Try running `vtex setup --typings` in the root folder for this repo, it should download the correct type definitions for all VTEX IO apps.

**Review by victorhmp — CHANGES_REQUESTED** [review]
Just a couple more changes.

Also, could you provide a better description and motivation for this change, along with a workspace where this can be tested? This makes it easier for us to test and also for our technical writers to write about this release in the next Release Notes :)

File: react/interfaces/custom-pixel-event.interface.ts (hunk 3 above)
Comment (igorbrasileiro) [reply react/interfaces/custom-pixel-event.interface.ts:1]: Also, this interface can be written like this

```
export type CustomPixelEventInterface = Record<string, CustomPixelEventPropTypes | CustomPixelEventPropTypes[]>
```

File: react/modules/customPixelEvents.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+  window.postMessage(data, origin)
+}
+
+/**
+ * Fires custom pixel event for newsletter
+ *
+ * @param {CustomPixelEventInterface} [eventData] - Event additional data
+ */
+export const triggerPixelNewsletterEvent = (
```
Comment (igorbrasileiro) [inline react/modules/customPixelEvents.ts:30]: These booth functions will be used in other apps?!

**Review by igorbrasileiro — COMMENTED** [review]
LGTM, but i didn't understand the purpose of this `triggerPixelEvent` and `triggerPixelNewsletterEvent` functions.

File: react/Newsletter.tsx (hunk 1 above)
Comment (igorbrasileiro) [reply react/Newsletter.tsx]: Remember to remove old functions

File: react/Newsletter.tsx [hunk 7]
```diff
@@ -19,14 +20,22 @@ interface Props {
   ErrorState?: ComponentType
   SuccessState?: ComponentType
   LoadingState?: ComponentType
+  CustomEventId?: string
```
[the remark this answers is not in the document]
Comment (Mogueno) [reply react/Newsletter.tsx]: @victorhmp Fixed.

File: react/interfaces/custom-pixel-event.interface.ts (hunk 3 above)
Comment (Mogueno) [reply react/interfaces/custom-pixel-event.interface.ts]: @victorhmp @igorbrasileiro yes, this was the old implementation, i`ve removed this file

File: react/modules/customPixelEvents.ts (hunk 4 above)
Comment (Mogueno) [reply react/modules/customPixelEvents.ts]: @victorhmp file removed

File: react/typings/vtex.pixel-manager.d.ts (hunk 5 above)
Comment (Mogueno) [reply react/typings/vtex.pixel-manager.d.ts]: @victorhmp Fixed typings by adding the dependency on manifest.json and running the command you suggested.

File: react/modules/customPixelEvents.ts (hunk 6 above)
Comment (Mogueno) [reply react/modules/customPixelEvents.ts]: @igorbrasileiro No, this is the old implementation, i`ve removed the files

**Comment by Mogueno** [comment]
> LGTM, but i didn't understand the purpose of this `triggerPixelEvent` and `triggerPixelNewsletterEvent` functions.

Those were the old implementation files, i`ve removed them

File: react/Newsletter.tsx [hunk 8]
```diff
@@ -109,14 +110,14 @@ function Newsletter(props: PropsWithChildren<Props>) {
       phone,
     }
 
-    const pixelEvent = customEventId
+    const pixelEvent: PixelData = customEventId
       ? {
           id: customEventId,
-          event: 'vtex:newsletterSubscription',
+          event: 'newsletterSubscription',
```
Comment (Mogueno) [inline react/Newsletter.tsx:116]: @victorhmp This event type is not yet implemented, that`s why the PixelData type is acusing an error on this const.

I`ve already sent this PR to fix that.
https://github.com/vtex-apps/pixel-app-template/pull/15

**Review by Mogueno — COMMENTED** [review]
> Just a couple more changes.
> 
> Also, could you provide a better description and motivation for this change, along with a workspace where this can be tested? This makes it easier for us to test and also for our technical writers to write about this release in the next Release Notes :)

Sure @victorhmp, I`ve never done this so correct me if I`m wrong - I need to provide you access to our development account for you to validate a workspace with this feature, right?

File: react/Newsletter.tsx (hunk 8 above)
Comment (icazevedo) [reply react/Newsletter.tsx:116]: @Mogueno if you send us the link to the workspace you used to test and validate the feature, we can access it. If it's a b2b store, it's desirable that you provide us credentials or help us create our user.

Comment (icazevedo) [reply react/Newsletter.tsx:116]: Also, you should rebase your code

**Comment by klzns** [comment]
Thanks for the pull request! Rebased and moved to https://github.com/vtex-apps/store-newsletter/pull/7

**Pull request closed without merging**